AR-1708 - Adding robustness to tests for incremental changes

crgee opened the pull request:
https://indigobio.atlassian.net/browse/AR-1708

Finished in 0.01854 seconds (files took 0.20523 seconds to load)
13 examples, 0 failures

wintonpc commented on lib/logatron/basic_logger.rb:
```diff
@@ -22,6 +22,10 @@ def debug(msg)
     end
 
+    def invalid_user(msg)
+      write(format_log(msg: msg, severity: INVALID_USE), SEVERITY_MAP[INVALID_USE])
```
is this really supposed to be `INVALID_USE` and not `INVALID_USER`?

jwliechty replied:
Or the reverse? Maybe this method is supposed to be called 

```
def invalid_use(msg)
```

which would make more sense. Either way, there is a typo.

crgee replied:
`INVALID_USE`

crgee replied:
Don't write code at night kids

jwliechty commented on lib/logatron/basic_scoped_logger.rb:
```diff
@@ -31,6 +31,10 @@ def debug(msg)
 
+    def invalid_user(msg)
+      write(format_log(msg: msg, severity: INVALID_USE), SEVERITY_MAP[INVALID_USE])
+    end
```
Again, is this method supposed to be called 

```
def invalid_use(msg)
```

instead?

wintonpc commented:
Why are we adding this new severity? The JIRA case is near-worthless.

jwliechty commented:
Cody mentioned to me that it is for logging exceptions that are normal workflow exceptions (such as validation). Does Justin need this severity when filtering in sumo-logic, or would logging at the `INFO` level had been adequate instead?

wintonpc commented:
This change requires a version bump.

wintonpc commented:
The readme lists the severities but does not explain that [the order is significant](https://github.com/indigobio/logatron/blob/new-severities/lib/logatron/const.rb#L12-L20).

Why is INVALID_USE lower severity than INFO ("successful operations, etc.")? It's not clear to me that it shouldn't be the other way around.

wintonpc commented:
`Logatron::Configuration#loggable_levels` is now broken because `Logatron::Configuration#new` wasn't updated for `INVALID_USE`. Rewrite `#new` to use `Logatron::SEVERITY_MAP` to eliminate the possibility of such a bug when severity levels are updated in the future.

Update spec `'Sets reasonable defaults'` for `INVALID_USE`

wintonpc commented:
The spec `'supports logging of various severities'` also needs to be updated for `INVALID_USE`

crgee commented:
The new severity is to catalog user behaviors. It would not be useful to use `INFO` unless you intend to filter out every `INFO` message that is not a user behavior. The severity of the issue is similar to `INFO` but it appears it needs to be ranked higher than `INFO` to make it into the logs. I am bumping the severity to 2 because of that, leaving `INFO` at 1.

When I made the initial changes I was given nearly no feedback to the impact of the changes at the time. (See prior pull request)

Given I was out of the office I appreciate the comments and am making quite a few changes to ensure any other additions to the logging severities causes failing tests.

wintonpc commented:
The previous commit shifted the integer values associated with the various levels in `SEVERITY_MAP`. Unfortunately, these values are [passed directly to ruby's logger](https://github.com/indigobio/logatron/blob/new-severities/lib/logatron/basic_logger.rb#L56-56), so now ruby's logger is receiving incorrect severities. (And in fact, it was already broken before, since ruby doesn't have a `CRITICAL` level.)

Confusingly, the spec `'supports logging of various severities'` still passes. If the severity passed to ruby's logger truly doesn't matter, then we should always pass the same value and document why in a comment. (Because Logatron already filters them based on severity before they get to the ruby logger, and the ruby logger defaults to DEBUG, which lets everything through. (??)) Or maybe avoid Ruby's logger altogether, if we're just treating it like an output stream.

Also, `Logatron::BasicLogger#write` should probably be private since it accepts a `SEVERITY_MAP` integer value, which is ostensibly an implementation detail. If any Logatron consumers call `write` directly, then such value-shifting changes will likely break them.

wintonpc commented:
The spec `'it supports a set of severities for logging'` also needs to be updated for `INVALID_USE`

wintonpc commented:
I would expect there to be a test that the logger level acts as a filter on the messages logged (i.e., drop the message if the message level is below the logger level), but I can't find one.

Commit:
AR-1708 - Adding invalid_use to tests, loggers

Finished in 0.01086 seconds (files took 0.11872 seconds to load)
13 examples, 0 failures

crgee commented:
@wintonpc severity # does seem to be related to the ruby logger. I'm not certain about the `CRITICAL` severity in that relationship. I set `INVALID_USE` to 1 (matching `INFO`).

I cannot set the `write` function to private - it causes tests to fail. I am going to defer any other changes for now to @djtreder or @grimesjm as they're the original authors and can attest to desired behavior.

Commit:
AR-1708 - Typo

wintonpc commented on lib/logatron/basic_scoped_logger.rb:
```diff
@@ -9,14 +9,7 @@ class BasicScopedLogger
-          CRITICAL => [],
-          FATAL => []
-      }
+      @logs = Hash[Logatron::SEVERITY_MAP.keys.map{|key| [key, []]}]
```
Good catch

wintonpc replied:
Good catch

wintonpc commented on lib/logatron/configuration.rb:
```diff
@@ -30,12 +30,7 @@ def initialize
-                ERROR,
-                CRITICAL,
-                FATAL]
+      levels = Logatron::SEVERITY_MAP.keys
```
Better, but this still unnecessarily assumes that the map elements are defined in the same order as their values. Filter based on `SEVERITY_MAP` values instead of the `Array#slice` hack below.

crgee replied:
Good point

wintonpc commented on lib/logatron/const.rb:
```diff
@@ -1,15 +1,19 @@
   SEVERITY_MAP =  {
       Logatron::DEBUG => 0,
       Logatron::INFO => 1,
+      Logatron::INVALID_USE => 1,
```
Having two with the same value warrants a comment.

crgee replied:
I'll make an entry into the readme - now that we know how the integer is used I can better articulate it

wintonpc commented on spec/lib/logatron/basic_logger_spec.rb:
```diff
@@ -20,7 +20,7 @@ module Logatron
         duration = 0
         Logatron.configure {|config| config.transformer = proc {|x| duration = x[:duration]; x[:duration].to_s}; config.logger = Logger.new(io); }
-        bl.log do |sl|
+        bl.log do |_|
```
just omit the parameter list altogether.

`bl.log { }`

wintonpc commented on spec/lib/logatron/basic_scoped_logger_spec.rb:
```diff
@@ -4,8 +4,8 @@
-      %w(info warn debug error critical fatal).each do |severity|
-        expect(BasicScopedLogger.method_defined? severity).to be(true)
+      Logatron::SEVERITY_MAP.keys.each do |severity|
+        expect(BasicScopedLogger.method_defined? severity.downcase).to be(true)
```
I would prefer to see all such cases factored as

`Logatron::SEVERITY_MAP.keys.map(&:downcase).each do |method_name|`

Separation of concerns.

Commit:
AR-1708 - Robust filtering of loggable levels, adding comments

Commit:
AR-1708 - Flipped conditional in my review, didn't run tests. Failed us all

Finished in 0.0138 seconds (files took 0.11666 seconds to load)
13 examples, 0 failures

wintonpc commented:
@crgee Go ahead and merge this once Jenkins gets his act together.

crgee commented:
retest this please
